aws-lc-rs/src/agreement.rs

```diff
@@ -530,25 +575,90 @@ where
+        }
     }
 
     let mut buffer = [0u8; MAX_AGREEMENT_SECRET_LEN];
 
     let secret: &[u8] = match &my_private_key.inner_key {
-        KeyInner::X25519(priv_key) => {
-            x25519_diffie_hellman(&mut buffer, priv_key, peer_pub_bytes).or(Err(error_value))?
-        }
-        KeyInner::ECDH_P256(priv_key)
-        | KeyInner::ECDH_P384(priv_key)
-        | KeyInner::ECDH_P521(priv_key) => {
-            ec_key_ecdh(&mut buffer, priv_key, peer_pub_bytes, expected_nid).or(Err(error_value))?
+        KeyInner::X25519(priv_key, ..) => match peer_public_key.encoding.id {
+            EncodingID::OctetString => {
+                x25519_dh_with_octstr_peer_pubkey(
+                    peer_pub_bytes,
+                    &mut buffer,
+                    priv_key,
+                    error_value,
+                )?;
+                &buffer[0..X25519_SHARED_KEY_LEN]
+            }
+            EncodingID::X509 => {
+                x25519_dh_with_x509_peer_pubkey(
+                    peer_pub_bytes,
+                    &mut buffer,
+                    priv_key,
+                    error_value,
+                )?;
+                &buffer[0..X25519_SHARED_KEY_LEN]
+            }
+        },
+        KeyInner::ECDH_P256(ec_key) | KeyInner::ECDH_P384(ec_key) | KeyInner::ECDH_P521(ec_key) => {
+            let pub_key_bytes = peer_public_key.bytes.as_ref();
+            ec_key_ecdh(
+                &mut buffer,
+                ec_key,
+                pub_key_bytes,
+                peer_public_key.encoding,
+                expected_nid,
+            )
+            .or(Err(error_value))?
         }
     };
     kdf(secret)
 }
 
+#[inline]
+fn x25519_dh_with_octstr_peer_pubkey<E>(
+    peer_pub_bytes: &[u8],
+    shared_secret: &mut [u8; MAX_AGREEMENT_SECRET_LEN],
+    priv_key: &LcPtr<EVP_PKEY>,
+    error_value: E,
+) -> Result<(), E> {
+    let mut pub_key = [0u8; X25519_PUBLIC_VALUE_LEN as usize];
+    pub_key.copy_from_slice(peer_pub_bytes);
+    x25519_diffie_hellman(shared_secret, priv_key, &pub_key).map_err(|()| error_value)?;
+    Ok(())
+}
+
+#[inline]
+fn x25519_dh_with_x509_peer_pubkey<E: Copy>(
+    peer_x509_pubkey: &[u8],
+    shared_secret: &mut [u8; MAX_AGREEMENT_SECRET_LEN],
+    priv_key: &LcPtr<EVP_PKEY>,
+    error_value: E,
+) -> Result<(), E> {
+    let len = match c_long::try_from(peer_x509_pubkey.len()) {
+        Ok(len) => len,
+        Err(_) => return Err(error_value),
+    };
+    let x509_pubkey = LcPtr::new(unsafe {
+        d2i_X509_PUBKEY(
+            null_mut(),
+            &peer_x509_pubkey.as_ptr() as *const *const u8 as *mut *const c_uchar,
+            len,
+        )
+    })
+    .map_err(|()| error_value)?;
+    let mut peer_octstr_pubkey = [0u8; PUBLIC_KEY_MAX_LEN];
+    marshal_x509_public_key_to_buffer(&mut peer_octstr_pubkey, &x509_pubkey)
+        .map_err(|_| error_value)?;
+    x25519_diffie_hellman_x509(shared_secret, priv_key, &peer_octstr_pubkey)
```

Comment:
Nit, maybe this should call x25519_dh_with_octstr_peer_pubkey rather than copying the end of that function here, now that we have the peer_octstr_pubkey?

It's interesting that we have X25519_PUBLIC_VALUE_LEN as the local buffer length vs. PUBLIC_KEY_MAX_LEN here -- I would have expected those to match unless I'm missing something?

Reply:
>Nit, maybe this should call x25519_dh_with_octstr_peer_pubkey rather than copying the end of that function here, now that we have the peer_octstr_pubkey?

This is not a nit.  A great catch!